models/PyCryptoBot.py

```diff
@@ -397,54 +399,63 @@ def getSmartSwitchDataFrame(self, df: pd.DataFrame, market, granularity: int, si
+                df1 = self.getHistoricalData(market, granularity,
+                                                str(df_first.isoformat()), 
+                                                str(simend.isoformat()))
+																								  
                 result_df_cache = df1
-                
-                df_first = self.getDateFromISO8601Str(datetime.strptime(str(simstart), "%Y-%m-%d %H:%M:%S").isoformat())
-                df_last = self.getDateFromISO8601Str(datetime.strptime(str(simend), "%Y-%m-%d %H:%M:%S").isoformat())
-                end_date = self.getDateFromISO8601Str(datetime.strptime(str(simend), "%Y-%m-%d %H:%M:%S").isoformat())
-                
-                if df_first.isoformat(timespec='milliseconds') > simstart.isoformat(timespec='milliseconds'):
-                    df1 = self.getHistoricalData(market, granularity, datetime.strptime(str(df_first), "%Y-%m-%d %H:%M:%S").isoformat(), datetime.strptime(str(end_date), "%Y-%m-%d %H:%M:%S").isoformat())
 
-                    while df_first.isoformat(timespec='milliseconds') > simstart.isoformat(timespec='milliseconds'):
-                        end_date = df_first
-                        df_first -= timedelta(minutes=(300*(granularity/60)))
+                while df_first.isoformat(timespec='milliseconds') > simstart.isoformat(timespec='milliseconds'):
+                    end_date = df_first
+                    df_first -= timedelta(minutes=(300*(granularity/60)))
                         
-                        if df_first.isoformat(timespec='milliseconds') < simstart.isoformat(timespec='milliseconds'):
-                                df_first = simstart
+                    if df_first.isoformat(timespec='milliseconds') < simstart.isoformat(timespec='milliseconds'):
+                        df_first = self.getDateFromISO8601Str(str(simstart))
+
+                    df2 = self.getHistoricalData(market, granularity, str(df_first.isoformat()), str(end_date.isoformat()))
 
-                        df2 = self.getHistoricalData(market, granularity, str(df_first), str(end_date))
+                    result_df_cache = pd.concat([df2.copy(), df1.copy()]).drop_duplicates()
+                    df1 = result_df_cache
 
-                        result_df_cache = pd.concat([df2.copy(), df1.copy()]).drop_duplicates()
-                        df1 = result_df_cache
-            if 'morning_star' not in result_df_cache:
+											  
+            if len(result_df_cache) > 0 and 'morning_star' not in result_df_cache:
+									  
                 result_df_cache.sort_values(by=['date'], ascending=True, inplace=True)
                 trading_dataCopy = result_df_cache.copy()
                 technical_analysis = TechnicalAnalysis(trading_dataCopy)
                 technical_analysis.addAll()
-                result_df_cache = technical_analysis.getDataFrame()
+                result_df_cache = trading_dataCopy
```

Comment:
why do we add TA indicators and then return the original frame without them?

Reply:
This happened after a refactor, I tried to remove the technical_analysis from here but when I did I started getting copy warnings from Pandas, I left it in for now with the intention of returning and trying to understand why it would do that. (it is not required in this function anymore)